### ISSUE TITLE
Get access to jwt, knoraApiConfig and knoraApiConnection in app or module

### ISSUE THREAD

**Issue by kilchenmann** [dasch-swiss/knora-api-js-lib#104]
We should store the json webtoken in browser's localstorage (or probably there are better places). Because right now, when we reload the app, the jwt info disappears and the user is not logged-in anymore. In the @knora/authentication I store the session with following information: 

- session id = timestamp
- user
  - username: string
  - jwt: string JSON Webtoken
  - lang: string User's language
  - sysAdmin: boolean Is the user system admin?
  - projectAdmin: string[] Array with project IRIs where the user is project admin

@tobiasschweizer We can have a look on that in detail and discuss the best solution.

**Comment by tobiasschweizer** [comment]
I think in this case the json webtoken has to be reset by the app using the endpoint's setter:

https://github.com/dasch-swiss/knora-api-js-lib/blob/84d7dddd9b1e8e4c2034f04056c265f0475f843b/src/api/endpoint.ts#L46-L48

**Comment by kilchenmann** [comment]
This works for me. Thanks.

**Comment by andreas-aeschlimann** [comment]
When the page is reloaded, I assume that you recreate the whole instances as well.

You can pass the JWT in the `KnoraApiConfig` instance that you create before instantiating `KnoraApiConnection`.

It works through the setter as well as @tobiasschweizer mentioned, but it makes less sense imho.

**Comment by kilchenmann** [comment]
> You can pass the JWT in the KnoraApiConfig instance that you create before instantiating KnoraApiConnection.

This happens in the AppInit service and I don't have access to this service from the knora-ui module. This is a bit an issue. @andreas-aeschlimann I tried your suggested solution but I wasn't able to set again the KnoraApiConfig.

**Comment by kilchenmann** [comment]
To understand the problem:

We have to use an AppInitService in the app:

```typescript
import { Injectable } from '@angular/core';
import { KnoraApiConfig, KnoraApiConnection } from '@knora/api';
import { KuiConfig } from '@knora/core';

@Injectable()
export class AppInitService {

    static knoraApiConnection: KnoraApiConnection;

    static knoraApiConfig: KnoraApiConfig;

    static kuiConfig: KuiConfig;

    constructor() { }

    Init() {

        return new Promise<void>((resolve, reject) => {

            // init knora-ui configuration
            AppInitService.kuiConfig = window['tempConfigStorage'] as KuiConfig;

            // init knora-api configuration
            AppInitService.knoraApiConfig = new KnoraApiConfig(
                AppInitService.kuiConfig.knora.apiProtocol,
                AppInitService.kuiConfig.knora.apiHost,
                AppInitService.kuiConfig.knora.apiPort
            );

            // set knora-api connection configuration
            AppInitService.knoraApiConnection = new KnoraApiConnection(AppInitService.knoraApiConfig);

            resolve();
        });
    }
}
```

To have access to the knoraApiConfig or the knoraApiConnection I have to define it in the app.module

```typescript
 providers: [
        AppInitService,
        {
            provide: APP_INITIALIZER,
            useFactory: initializeApp,
            deps: [AppInitService],
            multi: true
        },
        {
            provide: KuiConfigToken,
            useFactory: () => AppInitService.kuiConfig
        },
        {
            provide: KnoraApiConfigToken,
            useFactory: () => AppInitService.knoraApiConfig
        },
        {
            provide: KnoraApiConnectionToken,
            useFactory: () => AppInitService.knoraApiConnection
        }
    ],
```

and to inject it in the constructor, where I want to use one of the knora-api-js-lib service with:

```typescript
@Inject(KnoraApiConnectionToken) private knoraApiConnection: KnoraApiConnection
```

We did this in all our apps like this. It's needed because of the docker builds and the configuration (e.g. api url) from outside. 
@andreas-aeschlimann probably you have a better solution?

**Comment by andreas-aeschlimann** [comment]
I see. My thoughts about this:

1) The Config instance is indeed created at app load, however, it is passed by reference into the KnoraApiConnection. Thus, you can set `appInitService.knoraApiConfig.jsonWebToken = "asdf"`. It has the same effect as `knoraApiConnection.v2.auth.jsonWebToken = "asdf";`. To be on the safe side, make all properties in the InitService readonly. If you do this, you will be sure to have always access to the same instance (reference).

2) This has nothing to do with the issue: In the `AppInitService`, don't use static properties. It is against Angular philosophy. 

Suggestion:

```
import { Injectable } from '@angular/core';
import { KnoraApiConfig, KnoraApiConnection } from '@knora/api';
import { KuiConfig } from '@knora/core';

@Injectable()
export class AppInitService {

    // We cannot make readonly properties because we assign them later, so make properties that are only accessible in this class!
    private _knoraApiConnection: KnoraApiConnection;
    get knoraApiConnection(): KnoraApiConnection() { return this._knoraApiConnection; }

    private _knoraApiConfig: KnoraApiConfig;
    get knoraApiConnection(): KnoraApiConfig() { return this._knoraApiConfig; }

    private _kuiConfig: KuiConfig;
    get knoraApiConnection(): KuiConfig() { return this._kuiConfig; }

    constructor() { }

    // small i, TypeScript convention
    init() {

        return new Promise<void>((resolve, reject) => {

            // init knora-ui configuration
            this._kuiConfig = window['tempConfigStorage'] as KuiConfig;
            // Consider the use of json2typescript here, it can throw errors for you if there is a config issue

            // init knora-api configuration
            this._knoraApiConfig = new KnoraApiConfig(
                this._kuiConfig.knora.apiProtocol,
                this._kuiConfig.knora.apiHost,
                this._kuiConfig.knora.apiPort
            );

            // set knora-api connection configuration
            this._knoraApiConnection = new KnoraApiConnection(this._knoraApiConfig);

            resolve();
        });
    }
}
```

Like this, you can safely set `appInitService.knoraApiConfig.jsonWebToken = "asdf"` anywhere you like.

**Comment by kilchenmann** [comment]
I will try this

**Comment by kilchenmann** [comment]
> Like this, you can safely set appInitService.knoraApiConfig.jsonWebToken = "asdf" anywhere you like.

Not sure if I understood correct, but I don't think I'll have access to the appInitService outside of the app, e.g. in one of the knora-ui modules. I tried to move this appInitService into the core module, but I have to load the service before loading the app and this is not possible because we can't run a service of a module that is not loaded yet. I have to use configTokens as I described in app.module above. But with your suggestion I'm not able to set these configTokens?
